SAP_APIv1/SAP_API/Services/BakingProgramService.cs

```diff
@@ -11,152 +12,466 @@ public class BakingProgramService : IBakingProgramService
+            return numberOfProductsThatCanBeArranged;
+        }
+
+        private void FinishArrangingCurrentProgram(List<BakingProgram> bakingPrograms)
+        {
+            bakingPrograms.RemoveAt(0);
+        }
+
+        private void FinishArrangingCurrentProduct(List<OrderProduct> products)
+        {
+            products.RemoveAt(0);
+        }
+
+        private void ArrangeProductToProgram(OrderProduct product, int numberOfProductsToArrange, BakingProgram program, List<BakingProgram> programsProductsShouldBeArrangedTo)
+        {
+            programsProductsShouldBeArrangedTo.Remove(program);
+            BakingProgramProduct programProduct = new BakingProgramProduct
+            {
+                Id = new Guid(),
+                Product = product.Product,
+                Order = null,
+                QuantityТoBake = numberOfProductsToArrange
+            };
+            program.AddProductToProgram(programProduct);
+            product.QuantityToBake -= numberOfProductsToArrange;
+            programsProductsShouldBeArrangedTo.Add(program);
+        }
+        #endregion
+
+        #region arragingToNewPrograms
 
-                if(productIndex < productCount)
+        /// <summary>
+        /// To avoid complicated logic, when looking for free time slots for baking, 
+        /// the length of the time slot corresponds to the length of the longest baking time of 
+        /// the products from the set of products in the order
+        /// </summary>
+        /// <param name="groupedProductsDict"></param>
+        /// <returns></returns>
+        private List<BakingProgram> GetNewProgramsProductsShouldBeArrangedInto(Dictionary<TimeAndTempGroup, List<OrderProduct>> groupedProductsDict)
+        {
+            List<BakingProgram> programsProductsShouldBeArrangedTo = new List<BakingProgram>();
+            IEnumerable<TimeAndTempGroup> bakingGroups = groupedProductsDict.Keys;
+
+            int longestBakingTimeInMins = GetLongestBakingTimeForUnarrangedProducts(groupedProductsDict);
+            List<BakingProgram> bakingPrograms = GetAvailableProgramsForOvens(longestBakingTimeInMins);
+
+            foreach (TimeAndTempGroup group in bakingGroups)
+            {
+                List<OrderProduct> productsFromGroup = groupedProductsDict[group];
+                ArrangeProductsFromGroupToNewPrograms(bakingPrograms, productsFromGroup, group, programsProductsShouldBeArrangedTo);
+
+            }
+            return programsProductsShouldBeArrangedTo;
+        }
+
+        private int GetLongestBakingTimeForUnarrangedProducts(Dictionary<TimeAndTempGroup, List<OrderProduct>> groupedProductsDict)
+        {
+            int longestBakingTime = 0;
+            IEnumerable<TimeAndTempGroup> groups = groupedProductsDict.Keys;
+            foreach (TimeAndTempGroup group in groups)
+            {
+                List<OrderProduct> productsFromGroup = groupedProductsDict[group];
```

Comment:
Da nema slucajno lista neku sort funckiju gde prosledis kriterijum? Ako ima mozemo samo preko te funckije da sortuejmo i odaberemo prvog ili poslednjeg.

Primer na sta mislim :
```
productsFromGroup.sort((productA, productB) => productA.BakingTimeInMins> productB.BakingTimeInMins))
```

P.S ovo je presvetli JS

Reply:
Ovo verovatno sigurno postoji. Ovo ako si mislio na GetLongestBakingTimeForUnarrangedProducts -> trenutno bi problem bio jer u recniku mogu da ostanu TimeAndTempGroup grupe proizvoda u kojima su svi proizvodi prethodno dodeljeni postojecim programima (zato se proverava productsFromGroup.Count != 0). Jedino sto ne mora da se radi je da se prolazi kroz sve proizvode u jednoj grupi, to ne znam sto sam radila jer svi oni imaju jedno isto vreme:

  foreach (TimeAndTempGroup group in groups)
            {
                List<OrderProduct> productsFromGroup = groupedProductsDict[group];
                if (productsFromGroup.Count != 0)
                {
                    int bakingTimeForProducts = group.Time;
                    if (longestBakingTime < bakingTimeForProducts)
                        longestBakingTime = bakingTimeForProducts;

                }
            }